### ISSUE TITLE
Improve Output vs Input/Output Tests

### ISSUE THREAD

**Issue by niznik-dev** [niznik-dev/predicting-zygosity#35]
# Describe the Feature

We identified a few things to continue exploring after closing #17:

- Can we increase the GPU utilization to closer to 100%?
- Can we figure out what's going on with the "spikiness"?
  - Idea 1: Turn off or minimize the val loss calculation and see if that changes anything
  - Idea 2: Increase the length of the strings

## Related Issues

#17 

# Acceptance Criteria
- [ ] predictable_or_not can toggle and/or change the frequency of val loss calculation
- [ ] predictable_or_not has a parameter for setting how long the input sequence is

**Comment by niznik-dev** [comment]
@msalganik Can you add anything I'm missing? Feel free to edit the issue or reply with comments and I can organize it.

**Comment by niznik-dev** [comment]
Changing max_seq_len from 2048 to 2100 resulted in this, so it does seem sensitive to how packing is using that parameter (I tried to increase it just enough to make 10 even divisions per epoch and was pretty close!) ![Image](https://github.com/user-attachments/assets/5e47cdb5-0df4-45af-8707-12b05be9e9c0)

**Comment by niznik-dev** [comment]
Single epoch is stable after spinup

<img width="821" alt="Image" src="https://github.com/user-attachments/assets/61175087-c5b2-4f4b-a7b5-4451c5b70cb0" />

**Comment by msalganik** [comment]
Great.  I think this is resolved, but let's try to write an end to this story.

How would you summarize what we learned?  

Then we could decide if this should be a blog post or not.

**Comment by niznik-dev** [comment]
I need to perform a couple extra runs to see if saving and/or validation really does cause extra spikes (currently not 💯 sure) but I will paste my summary when I have those runs.

**Comment by niznik-dev** [comment]
This image helped me confirm some of what I was still curious about. (1) You'll notice that the purple run where saving is turned off appears to only have a one step drop in tokens per second (the "short" packed grouping) at the end of the first few epochs compared to the green one which is saving and sees more like a two step drop (the "short" packed grouping and then the saving artifact). (2) I ran validation with a clear offset here (40 steps) which you can clearly see because the step count will skip from 39 to 41, etc. Even so, you see another spike due to resuming work after validation being a bit slow to resume. So regardless of what we found about the packing first, shuffling second done per epoch, it appears saving and validation cause spikes as well. So we have three total spike reasons ("short" packed groupings, end-of-epoch saving, and validation) which may drop to 2 when max_seq_len is chosen thoughfully...or serendipitously.

<img width="1272" alt="Image" src="https://github.com/user-attachments/assets/80c52b00-3195-4f24-a08c-58431bbfce4e" />

**Comment by niznik-dev** [comment]
My attempt at a summary:

We began this exercise in #17 to try and figure out (a) why our tokens per second seemed to be on the low side when showing results to colleagues and (b) a sanity check on us understanding how the model performs in both scenarios given different combinations of predictable and unpredictable outputs. Matt and I made a series of predictions, but after squashing some bugs it appeared that Matt's predictions were spot on. In addition, we confirmed that our tokens per second counts were low primarily because the input, by default, was being converted into masked tokens which were not counted in tokens per second. This is fine as long as everyone is aware of this (and we can in theory calculate a metric like "observed/processed tokens per second" to distinguish from "finetuned tokens per second" if needed).

A question raised was "why are the tokens spiky in the W&B graphs"? This moved to #35 for further examination. In the process, we learned that this was due to a combination of (a) the order of packing/shuffling (packing first, then shuffling the packed groups, one of which will usually be smaller than the rest but padded with masked tokens up to max_seq_len), (b) disabling saving and validation loss calculations further smoothed things out. Because these are primarily "end of epoch" issues (with shuffling disabled), we are now relatively unconcerned with our spikes.
